url6544f14c72_charles_d_burton_java_websocket_example_tgz-pJ8-minewebsocket_MineWebSocketJ8/src/minewebsocket/org/java_websocket/WebSocketImpl.java

```diff
@@ -0,0 +1,737 @@
+									}
+									write( d.createHandshake( d.postProcessHandshakeResponseAsServer( handshake, response ), role ) );
+									draft = d;
+									open( handshake );
+									return true;
+								}
+							} catch ( InvalidHandshakeException e ) {
+								// go on with an other draft
+							}
+						}
+						if( draft == null ) {
+							close( CloseFrame.PROTOCOL_ERROR, "no draft matches" );
+						}
+						return false;
+					} else {
+						// special case for multiple step handshakes
+						Handshakedata tmphandshake = draft.translateHandshake( socketBuffer );
+						if( tmphandshake instanceof ClientHandshake == false ) {
+							flushAndClose( CloseFrame.PROTOCOL_ERROR, "wrong http function", false );
+							return false;
+						}
+						ClientHandshake handshake = (ClientHandshake) tmphandshake;
+						handshakestate = draft.acceptHandshakeAsServer( handshake );
+
+						if( handshakestate == HandshakeState.MATCHED ) {
+							open( handshake );
+							return true;
+						} else {
+							close( CloseFrame.PROTOCOL_ERROR, "the handshake did finaly not match" );
+						}
+						return false;
+					}
+				} else if( role == Role.CLIENT ) {
+					draft.setParseMode( role );
+					Handshakedata tmphandshake = draft.translateHandshake( socketBuffer );
+					if( tmphandshake instanceof ServerHandshake == false ) {
+						flushAndClose( CloseFrame.PROTOCOL_ERROR, "wrong http function", false );
+						return false;
+					}
+					ServerHandshake handshake = (ServerHandshake) tmphandshake;
+					handshakestate = draft.acceptHandshakeAsClient( handshakerequest, handshake );
+					if( handshakestate == HandshakeState.MATCHED ) {
+						try {
+							wsl.onWebsocketHandshakeReceivedAsClient( this, handshakerequest, handshake );
+						} catch ( InvalidDataException e ) {
+							flushAndClose( e.getCloseCode(), e.getMessage(), false );
+							return false;
+						} catch ( RuntimeException e ) {
+							wsl.onWebsocketError( this, e );
+							flushAndClose( CloseFrame.NEVER_CONNECTED, e.getMessage(), false );
+							return false;
+						}
+						open( handshake );
+						return true;
+					} else {
+						close( CloseFrame.PROTOCOL_ERROR, "draft " + draft + " refuses handshake" );
+					}
+				}
+			} catch ( InvalidHandshakeException e ) {
+				close( e );
+			}
+		} catch ( IncompleteHandshakeException e ) {
+			if( tmpHandshakeBytes.capacity() == 0 ) {
+				socketBuffer.reset();
+				int newsize = e.getPreferedSize();
+				if( newsize == 0 ) {
+					newsize = socketBuffer.capacity() + 16;
+				} else {
+					assert ( e.getPreferedSize() >= socketBuffer.remaining() );
+				}
+				tmpHandshakeBytes = ByteBuffer.allocate( newsize );
+
+				tmpHandshakeBytes.put( socketBufferNew );
+				// tmpHandshakeBytes.flip();
+			} else {
+				tmpHandshakeBytes.position( tmpHandshakeBytes.limit() );
+				tmpHandshakeBytes.limit( tmpHandshakeBytes.capacity() );
+			}
+		}
+		return false;
+	}
+
+	private void decodeFrames( ByteBuffer socketBuffer ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 23. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.